mono/metadata/marshal.c

```diff
@@ -3265,6 +3278,63 @@ mono_marshal_get_delegate_invoke_internal (MonoMethod *method, gboolean callvirt
 
 	mono_mb_emit_byte (mb, CEE_RET);
 
+	/* else [delegates != null] */
+	mono_mb_patch_branch (mb, pos3);
+
+	/* len = delegates.Length; */
+	mono_mb_emit_ldloc (mb, local_delegates);
+	mono_mb_emit_byte (mb, CEE_LDLEN);
+	mono_mb_emit_byte (mb, CEE_CONV_I4);
+	mono_mb_emit_stloc (mb, local_len);
+
+	/* if (len >= 2) */
+	mono_mb_emit_ldloc (mb, local_len);
+	mono_mb_emit_icon (mb, 2);
+	pos2 = mono_mb_emit_branch (mb, CEE_BLT);
+
+	/* int i = 0; */
+	mono_mb_emit_icon (mb, 0);
+	mono_mb_emit_stloc (mb, local_i);
+
+	pos1 = mono_mb_get_label (mb);
+
+	/* d = delegates [i]; */
+	mono_mb_emit_ldloc (mb, local_delegates);
+	mono_mb_emit_ldloc (mb, local_i);
+	mono_mb_emit_byte (mb, CEE_LDELEM_REF);
+	mono_mb_emit_stloc (mb, local_d);
+
+	/* res = d.Invoke ( args .. ); */
+	mono_mb_emit_ldloc (mb, local_d);
+	for (i = 0; i < sig->param_count; i++)
+		mono_mb_emit_ldarg (mb, i + 1);
+	if (!ctx) {
+		mono_mb_emit_op (mb, CEE_CALLVIRT, method);
+	} else {
+		MonoError error;
+		mono_mb_emit_op (mb, CEE_CALLVIRT, mono_class_inflate_generic_method_checked (method, &container->context, &error));
+		g_assert (mono_error_ok (&error)); /* FIXME don't swallow the error */
+	}
+	if (!void_ret)
+		mono_mb_emit_stloc (mb, local_res);
+
+	/* ++i < l */
+	mono_mb_emit_add_to_local (mb, local_i, 1);
+	mono_mb_emit_ldloc (mb, local_i);
+	mono_mb_emit_ldloc (mb, local_len);
+	mono_mb_emit_branch_label (mb, CEE_BLT, pos1);
+
+	/* return res */
+	if (!void_ret)
+		mono_mb_emit_ldloc (mb, local_res);
+	mono_mb_emit_byte (mb, CEE_RET);
+
+	/* else [len < 2] */
+	mono_mb_patch_branch (mb, pos2);
+
+	/* delegates should not be empty */
+	mono_mb_emit_exception_full (mb, "System", "InvalidOperationException", "delegates should be of size >= 2");
```

Comment:
it's more a protection against edge cases that can come bite us, and that can be pretty hard to debug. it's totally possible to remove.